aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -89,11 +93,29 @@ void copy_device_to_device(TensorIterator& iter, bool non_blocking) {
     void *src = iter.data_ptr(1);
     size_t size = numel * iter.element_size(0);
     if (src != dst || src_device != dst_device) {
-      // Perform the copy
-      AT_CUDA_CHECK(cudaMemcpyAsync(
-          dst, src, size,
-          cudaMemcpyDeviceToDevice,
-          copy_stream));
+#if CUDA_VERSION >= 11040
+      // Due to bizarre cuda driver intricacies, copies of
+      // cudaMallocAsynced memory between devices that aren't
+      // peer-to-peer-capable need "cudaMemcpyPeerAsync".
+      static bool using_cudaMallocAsync = std::strcmp(CUDACachingAllocator::allocatorBackend(),
+                                                      "cudaMallocAsync") == 0;
+      bool needs_MemcpyPeer = (src_device != dst_device &&
+                               using_cudaMallocAsync &&
+                               !p2p_enabled);
+      if (needs_MemcpyPeer) {
+        AT_CUDA_CHECK(cudaMemcpyPeerAsync(
+            dst, dst_device.index(),
+            src, src_device.index(),
+            size, copy_stream));
+      } else {
+#endif
+        AT_CUDA_CHECK(cudaMemcpyAsync(
+            dst, src, size,
+            cudaMemcpyDeviceToDevice,
+            copy_stream));
+#if CUDA_VERSION >= 11040
+      }
+#endif
```

Comment:
@ezyang per https://github.com/pytorch/pytorch/pull/65365/files#r822992141 i'll also avoid stubbing out a single curly here.

Reply:
Unlike calls in https://github.com/pytorch/pytorch/pull/65365/files#r822992141 the code here should be compileable with any cuda version, so no macros needed.